### PR TITLE
feat: auto tile view

### PR DESCRIPTION
**Pull request by zbettenbuk**
This PR implements auto tile view based on the following logic:

- Tile view now has an "undefined" state, that means that the user hasn't requested any views in particular. In this case we use auto switching based on a few rules (see `video-layout/functions.js`)
- Once the user has selected one, we try to respect that, however we still auto-switch to tile view on screen share or Etherpad editing (is that still a thing btw?)
- Auto switching respects the previously selected state when the cause for auto switch is over
- We don't persist the tile view setting anymore to promote auto switching

### CODE REVIEW

File: react/features/video-layout/reducer.js [hunk 1]
```diff
@@ -14,18 +14,17 @@ const DEFAULT_STATE = {
      * The indicator which determines whether the video layout should display
      * video thumbnails in a tiled layout.
      *
+     * Note: undefined means that the user hasn't requested anything in particular yet, so
+     * we use our auto shitching rules.
```
Comment (saghul) [inline react/features/video-layout/reducer.js:18]: switching

File: modules/UI/videolayout/SmallVideo.js [hunk 2]
```diff
@@ -451,7 +451,7 @@ export default class SmallVideo {
      */
     selectDisplayMode(input) {
         // Display name is always and only displayed when user is on the stage
-        if (input.isCurrentlyOnLargeVideo && !input.tileViewEnabled) {
+        if (input.isCurrentlyOnLargeVideo && !input.tileViewActive) {
```
Comment (saghul) [inline modules/UI/videolayout/SmallVideo.js:454]: Is this change necessary?

File: react/features/video-layout/middleware.any.js [hunk 3]
```diff
@@ -1,16 +1,15 @@
...
 import { MiddlewareRegistry } from '../base/redux';
-import { SET_DOCUMENT_EDITING_STATUS, toggleDocument } from '../etherpad';
+import { SET_DOCUMENT_EDITING_STATUS } from '../etherpad';
 
-import { SET_TILE_VIEW } from './actionTypes';
 import { setTileView } from './actions';
 
 import './subscriber';
 
+let previousTileViewEnabled;
```
[the remark this answers is not in the document]
Comment (zbettenbuk) [reply react/features/video-layout/middleware.any.js:12]: I had the thinking that this is strictly internal and temporary value and didn't want to do state updates for this

File: react/features/video-layout/functions.js [hunk 4]
```diff
@@ -72,17 +72,32 @@ export function getTileViewGridDimensions(state: Object, maxColumns: number = ge
...
+    }
+
+    // None tile view mode is easier to calculate (no need for many negations), so we do
+    // that and negate it only once.
+    const shouldDisplayNormalMode = Boolean(
+
+        // Reasons for normal mode:
+
+        // Editing etherpad
+        state['features/etherpad']?.editing
```
[the remark this answers is not in the document]
Comment (zbettenbuk) [reply react/features/video-layout/functions.js:90]: True, thanks!

File: react/features/toolbox/components/web/Toolbox.js [hunk 5]
```diff
@@ -58,7 +58,8 @@ import {
 } from '../../../subtitles';
 import {
     TileViewButton,
-    toggleTileView
+    toggleTileView,
+    shouldDisplayTileView
```
Comment (saghul) [inline react/features/toolbox/components/web/Toolbox.js:62]: alpha sort

File: react/features/video-layout/functions.js [hunk 6]
```diff
@@ -72,17 +73,44 @@ export function getTileViewGridDimensions(state: Object, maxColumns: number = ge
...
+
+        // We're in filmstrip-only mode
+        || (typeof interfaceConfig === 'object' && interfaceConfig?.filmStripOnly)
```
Comment (saghul) [inline react/features/video-layout/functions.js:103]: do we need the `typeof` check if we use optional chaining?

**Comment by saghul** [comment]
Hum, that test failure might be legit.

File: react/features/video-layout/functions.js (hunk 6 above)
Comment (zbettenbuk) [reply react/features/video-layout/functions.js:103]: Yes, it crashed otherwise :D

**Comment by crprata** [comment]
@saghul @zbettenbuk Hello! Is it possible do disable this auto tile view? Or at least to have the option to increase the number of participants for active the automatic tile view.
